web/mysql_typemap.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package web
+
+import (
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+// TODO(searce): Decide on the issue message that needs to be shown
+// in frontend.
+var mysqlTypeMap = map[string][]typeIssue{
```

Comment:
maybe I am missing some context here, but do we need to keep this map since we already have `toSpannerTypeMySQL` function  which gives the same mapping effectively?

similarly in web/postgresql_typemap.go

Reply:
yes Asheesh, we need to keep this. this map basically gives options from source type to possible spanner types, which will be sent to frontend. user will be able to select spanner type from possible options.
While toSpannerType function will just convert from source type to default spanner type.

Reply:
I'm also wondering about the substantial overlap between mysqlTypeMap and toSpannerTypeMySQL.
The main differences I see are:
a) the map just has a type id for the spanner type, whereas the function returns the full dd.Type.
b) the function handles mods (but the map doesn't)

I'm wondering if there is a way to combine them somehow?

Reply:
Maybe we can generate mysqlTypeMap from toSpannerTypeMySQL in a module initializer.

Something like:

func init() {
  for _, srcType := range []string{ /* list of all source types */ } {
       for _, spType := range []string { /* list of all spanner types */ } {
            var l []TypeIssue
           ty, issues := toSpannerTypeMySQL(conv, srcType, spType, []int64{})
           if ty.Name == spType {
              l = append(l, typeIssue{T:spType, Issue: issues} // Why does toSpannerTypeMySQL return a slice of issues, but typeIssue is just a single issue?
           }
       }
       mysqlTypeMap[srcType] = l
  }
}

I think that should be workable -- what do you think?

Reply:
yeah, this should work, i'll implement it, the reason why toSpannerTypeMySQL is returning array of issues is because we are creating a list of issues in the caller function just like original function toSpannerType.